Client/Frontend/Widgets/SnackBar.swift

```diff
@@ -0,0 +1,225 @@
+        return true
+    }
+
+    override func layoutSubviews() {
+        if let img = imageView.image {
+            imageView.snp_makeConstraints({ make in
+                make.left.equalTo(self).offset(DefaultPadding)
+                make.top.equalTo(self).offset(DefaultPadding)
+                make.width.equalTo(img.size.width)
+                make.height.equalTo(img.size.height)
+            })
+        } else {
+            imageView.snp_makeConstraints({ make in
+                make.width.equalTo(0)
+                make.height.equalTo(0)
+                make.top.left.equalTo(self).offset(DefaultPadding)
+            })
+        }
+
+        let labelSize = self.textView.sizeThatFits(CGSizeMake(self.frame.width, CGFloat(MAXFLOAT)))
+        textView.textContainerInset = UIEdgeInsetsZero
+        textView.snp_makeConstraints({ make in
+            make.top.equalTo(self.imageView.snp_top)
+            make.left.equalTo(self.imageView.snp_right)
+
+            make.height.equalTo(labelSize.height)
+            make.trailing.equalTo(self)
+        })
+
+        buttonsView.snp_makeConstraints({ make in
+            make.bottom.equalTo(self)
+            make.left.right.equalTo(self)
+            if self.buttonsView.subviews.count > 0 {
+            	make.height.equalTo(ToolbarHeight)
+            } else {
+                make.height.equalTo(0)
+            }
+        })
+
+        self.snp_makeConstraints({ make in
+            var h = labelSize.height
+            if let img = self.imageView.image {
+                h = max(2 * DefaultPadding + img.size.height, labelSize.height)
+            }
+
+            let constraint = make.height.equalTo(h)
+            if (self.buttonsView.subviews.count > 0) {
+                constraint.offset(ToolbarHeight)
+            }
+        })
+        super.layoutSubviews()
+    }
+
+    func show() {
+        alpha = 1
+        transform = CGAffineTransformIdentity
+    }
+
+    func hide() {
+        alpha = 0
+        var h = frame.height
+        if h == 0 {
+            h = ToolbarHeight
+        }
+        transform = CGAffineTransformMakeTranslation(0, h)
+    }
+
+    lazy var highlightImg: UIImage = {
+        let size = CGSize(width: 1, height: 1)
+        return UIImage.createWithColor(size, color: UIColor(red: 205/255, green: 223/255, blue: 243/255, alpha: 0.9))
+    }()
+
+    private func addButton(snackButton: SnackButton) {
+        let button = UIButton()
+        button.setTitle(snackButton.title, forState: .Normal)
+        button.titleLabel?.font = font
+        button.setBackgroundImage(highlightImg, forState: .Highlighted)
+        button.setTitleColor(UIColor(red: 42/255, green: 121/255, blue: 213/255, alpha: 1.0), forState: .Highlighted)
+
+        snackButton.bar = self
+        targets.append(snackButton) // We hold a reference to the target so that it isn't cleaned up
+        button.addTarget(snackButton, action: "onClick", forControlEvents: .TouchUpInside)
```

Comment:
If you make `SnackButton` a `UIControl` subclass, the `addTarget()` can move into the `SnackButton` code. I think that is where it belongs, since it is a local concern of the button to handle taps on it.